Allow the driver pod's credentials to be shipped from the submission client

mccheah opened the pull request:
Provide a separate set of configurations for loading Kubernetes API server credentials for the driver pod to use when requesting executors.

Credentials specified in the submitter are POST'd to the submission server and used by the driver for interactions with the apiserver.  This is a different set of credentials from those used by the submitter when starting the driver pod, allowing for fine-grained access controls between these two k8s clients if desired.

Closes #182

mccheah committed:
Allow the driver pod's credentials to be shipped through secrets.

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -200,42 +200,76 @@ from the other deployment modes. See the [configuration page](configuration.html
 </tr>
 <tr>
-  <td><code>spark.kubernetes.submit.caCertFile</code></td>
+  <td><code>spark.kubernetes.apiserver.submit.caCertFile</code></td>
```
Rather than introduce more k8s terminology, why don't we call this `spark.kubernetes.authentication.*`?

mccheah replied:
Done

ash211 commented:
Please rebase onto `branch-2.1-kubernetes` and send this PR into that branch instead of the `k8s-support-alternate-incremental` branch which is now deprecated.

mccheah committed:
Fix scalastyle

mccheah committed:
Change apiserver -> authentication

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -200,42 +200,76 @@ from the other deployment modes. See the [configuration page](configuration.html
+  <td><code>spark.kubernetes.authentication.driver.caCertFile</code></td>
+  <td>(none)</td>
+  <td>
+    CA cert file for connecting to Kubernetes over SSL from the driver pod when requesting executors. This file should
```
Kubernetes -> the Kubernetes API server

Start with "Path to the CA cert file" - and in other config options below

What URI scheme is expected, and what happens to paths with no scheme?

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/DriverPodKubernetesCredentialsProvider.scala:
```diff
@@ -0,0 +1,100 @@
+    sparkConf: SparkConf,
+    kubernetesAppId: String) {
+
+  def getDriverPodKubernetesCredentials(): DriverPodKubernetesCredentials = {
```
call this just `get` since it's pretty clear what you're getting from a `DriverPodKubernetesCredentialsProvider` and it's the only public method

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/DriverPodKubernetesCredentialsProvider.scala:
```diff
@@ -0,0 +1,100 @@
+        throw new SparkException(s"File provided for ${conf.key} at ${file.getAbsolutePath}" +
+          s" does not exist or is not a file.")
+      }
+      (secretName, BaseEncoding.base64().encode(Files.toByteArray(file)))
```
does `Files.toByteArray` close the `File` object?

mccheah replied:
The `File` object is not opened or closed, but rather the stream that pulls the data has to be opened and closed, which Guava takes care of.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/DriverPodKubernetesCredentialsProvider.scala:
```diff
@@ -0,0 +1,100 @@
+      .withMountPath(DRIVER_CONTAINER_KUBERNETES_CREDENTIALS_SECRETS_BASE_DIR)
+      .build()
+    // Cannot use both service account and mounted secrets
+    sparkConf.get(KUBERNETES_SERVICE_ACCOUNT_NAME).foreach { _ =>
```
move this validation up to right after these are created (validate early)

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/DriverPodKubernetesCredentialsProvider.scala:
```diff
@@ -0,0 +1,100 @@
+    val credentialsSecretVolumeMount = new VolumeMountBuilder()
+      .withName(credentialsSecretVolume.getName)
+      .withReadOnly(true)
+      .withMountPath(DRIVER_CONTAINER_KUBERNETES_CREDENTIALS_SECRETS_BASE_DIR)
```
I'm not totally sure I follow the mounting here -- why do we need to create a volume mount to store data to a file?  Is the docker image read-only or something?  Or we want to ensure the secrets are disposed of in a secure manner?

mccheah replied:
The docker image only provides a static state that the container starts with, and the volume mounts are attached to that container to add dynamic data from Volume plugins.

ash211 replied:
so this is a mechanism to inject files into the running container?

mccheah replied:
Not an already-running container exactly, but more precisely to attach one on startup of the container

mccheah replied:
We use this everywhere we've been providing secrets; the other way secrets can be consumed is via environment variables, but in all of our cases we've been loading them from mounted files.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesClientBuilder.scala:
```diff
@@ -25,29 +25,60 @@ import io.fabric8.kubernetes.client.{Config, ConfigBuilder, DefaultKubernetesCli
       .withApiVersion("v1")
       .withMasterUrl(KUBERNETES_MASTER_INTERNAL_URL)
       .withNamespace(kubernetesNamespace)
 
```
should we do the just-one-auth validation here too?

```
require((MOUNTED_TOKEN.isFile || MOUNTED_CA_CERT.isFile || MOUNTED_CLIENT_KEY.isFile || MOUNTED_CLIENT_CERT.isFile) ^
        (SERVICE_ACCOUNT_CA_CERT.isFile || SERVICE_ACCOUNT_TOKEN.isFile))
```

mccheah replied:
Think it's fine to leave it out here and rely on the client to check.

mccheah replied:
The service account token file is actually always present, so this check wouldn't be correct.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/SslConfigurationProvider.scala:
```diff
@@ -49,7 +49,8 @@ private[spark] class SslConfigurationProvider(
   private val sslSecretsName = s"$SUBMISSION_SSL_SECRETS_PREFIX-$kubernetesAppId"
-  private val sslSecretsDirectory = s"$DRIVER_CONTAINER_SECRETS_BASE_DIR/$kubernetesAppId-ssl"
+  private val sslSecretsDirectory = DRIVER_CONTAINER_SUBMISSION_SECRETS_BASE_DIR
+    s"/$kubernetesAppId-ssl"
```
does Scala magically concatenate these two adjacent strings?

mccheah replied:
Nope, this is a typo

mccheah replied:
I think this is failing the integration tests

ash211 replied:
but but but how does it compile?  does the hanging string just not get assigned to anything and disappear?

mccheah replied:
Believe that's the case yeah

ash211 commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesSuite.scala:
```diff
@@ -364,8 +308,12 @@ private[spark] class KubernetesSuite extends SparkFunSuite with BeforeAndAfter {
         .pods
         .withLabel("spark-app-name", "spark-file-existence-test")
         .watch(watch)) { _ =>
-      SparkSubmit.main(args)
```
why change from `SparkSubmit.main(args)` to `new Client(...)` ?

mccheah replied:
`SparkSubmit.main` is unwieldy because

(1) It's difficult to re-use arguments since they have to be formatted as an array rather than as a Map, and more importantly
(2) They deal with system properties which is global state across all tests in the JVM.

mccheah replied:
It's beneficial to test using `SparkSubmit` if we expect `SparkSubmit` to process its arguments in a way that our `Client` doesn't expect, but for the most part otherwise we should call the `Client` directly.

ash211 reviewed (commented):
what do you think of passing the creds through to the driver pod in the POST rather than via volume mount?  does it make a big difference one way or the other?

ash211 commented:
@kimoonkim heads up another flake in `SparkContextSuite`: http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-spark-k8s-unit-tests/25/consoleFull#-1164448660c3852bdb-3a63-45ca-a0d4-973026729059

mccheah commented:
Hm, I wanted to avoid making the submission server more complex by adding more data to the POST, and in the future if we switch to a different submission mechanism we'd have to rewrite more logic there. But I suppose creating fewer Kubernetes resources is beneficial as well particularly if these objects are short-lived.

ash211 commented:
retry the integration tests
retry the unit tests

ash211 commented:
rerun integration test please

ash211 commented:
rerun unit tests please

ash211 commented:
rerun integration test please

(previous failed: http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-spark-k8s-integration-test-/20/consoleFull#-167881559204b09c7b-0d94-4ce5-8a08-8f343248b3d8 )

ash211 commented:
rerun integration test please

mccheah commented:
Hm, should we also allow pre-mounted files, like we currently do for the keyStore? e.g. `local://` paths. I'm not sure how common it would be to bake Kubernetes credentials into docker images though, where it might be more common for TLS keys and certificates; the case here seems more user-specific while the TLS scenario could be fixed across many users of a Spark Docker image.

ash211 commented:
It seems reasonable to bake in a truststore for the apiserver into docker images, but I'm guessing most places would avoid putting creds into images for security reasons.

I'd propose we don't support `local://` paths for the various credentials on the driver until it affects someone.

mccheah committed:
Address comments.

Also some quality of life fixes, most notably formatting all of the
documentation strings in config.scala to no longer use triple quotes.
Triple quoted strings are difficult to format consistently.

mccheah committed:
Fix scalastyle

mccheah committed:
Fix comment

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:
```diff
@@ -27,233 +27,240 @@ package object config {
-          | this can also be passed to spark-submit via the
-          | --kubernetes-namespace command line argument.
-        """.stripMargin)
+      .doc("The namespace that will be used for running the driver and executor pods. When using" +
```
I reformatted all of these because it's difficult to get triple-quoting to have a consistent format.

mccheah replied:
This matches what YARN has.

mccheah committed:
Remove unnecessary constants

mccheah committed:
Remove unnecessary whitespace

ash211 commented:
rerun unit tests please

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -200,42 +200,88 @@ from the other deployment modes. See the [configuration page](configuration.html
+  <td>(none)</td>
+  <td>
+    Path to the client key file for authenticating against the Kubernetes API server from the driver pod when requesting
+    executors. This file should be located on the submitting machine's disk, and will be uploaded to the driver pod.
```
should -> must ?

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala:
```diff
@@ -210,7 +204,7 @@ private[spark] class Client(
         } finally {
           Utils.tryLogNonFatalError {
-            kubernetesResourceCleaner.deleteAllRegisteredResourcesFromKubernetes(kubernetesClient)
+//            kubernetesResourceCleaner.deleteAllRegisteredResourcesFromKubernetes(kubernetesClient)
```
was this line accidentally included?

mccheah replied:
Ah, yep

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesClientBuilder.scala:
```diff
@@ -22,64 +22,62 @@ import com.google.common.base.Charsets
+    val configBuilder = oauthTokenFile
+        .orElse(caCertFile)
+        .orElse(clientKeyFile)
+        .orElse(clientCertFile)
```
this seems like an unusual pattern for "if any of these are present".  Does Scala have an `any(Seq[Option[])` method like python's `any` ?

mccheah replied:
Don't see such an API, unfortunately.

ash211 replied:
Looks like the Scala way (from 2011 via http://stackoverflow.com/questions/6390797/scala-any-and-all-functions and 2013 via http://stackoverflow.com/questions/15932137/checking-that-all-items-in-a-collection-match-a-predicate-in-scala) is:

`Seq(oauthTokenFile, caCertFile, clientKeyFile, clientCertFile).any(_.isDefined)`

Do you find that clearer than the .orElse chain ?

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesClientBuilder.scala:
```diff
@@ -22,64 +22,62 @@ import com.google.common.base.Charsets
-      if (MOUNTED_CA_CERT.isFile) {
-        clientConfigBuilder = clientConfigBuilder.withCaCertFile(MOUNTED_CA_CERT.getAbsolutePath)
+      caCertFile.foreach { caFile =>
+        mountedAuthConfigBuilder = mountedAuthConfigBuilder.withCaCertFile(caFile)
```
this has to be reassigned every time?  would be nice if the fluent API allowed repeated calls on the same reference in these kinds of situations

mccheah replied:
Reading the Kubernetes client code indicates that we don't have to assign here - but I actually think it's clearer to follow this pattern currently, as it semantically means "This builder is equal to the old builder with these updates". The phrasing of the API (`builder.withCaCertFile`) made it sound like it doesn't mutate the underlying object - I would have expected `setCaCertFile` if it did.

ash211 reviewed (commented):
so the process is (using oauth token but same for the client cert pair):
- user sets oauth token in config key `spark.kubernetes.authentication.driver.oauthToken`
- submission client reads that out of SparkConf and creates a `KubernetesCredentials` out of it
- then redacts the token out of SparkConf
- posts the `KubernetesCredentials` to the submission server as part of job launch
- submission server then writes the token to disk in `writeKubernetesCredentials` and saves its path to SparkConf in the `spark.kubernetes.authentication.driver .mounted.oauthTokenFile` key
- driver then reads that path from its SparkConf and assembles a kubernetes client with it, in `buildFromWithinPod`

This seems pretty reasonable, and POST'ing those things directly to the driver pod vs the temporary volumes follows the pattern of other parts being posted.

No more major comments from me -- @foxish have you had a chance to give this a review?

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -200,42 +200,88 @@ from the other deployment modes. See the [configuration page](configuration.html
 </tr>
 <tr>
-  <td><code>spark.kubernetes.submit.caCertFile</code></td>
+  <td><code>spark.kubernetes.authentication.submission.caCertFile</code></td>
```
this key is kinda long -- thoughts on shortening `authentication` to just `auth` ?

mccheah replied:
They use `spark.authenticate` in other places - but not sure if we should be following that precedent: https://spark.apache.org/docs/latest/configuration.html

mccheah committed:
Authentication -> Authenticate

ash211 reviewed (approved):
LGTM!

@foxish any last comments?  I'm planning to merge this EOD if you don't have anything.

foxish commented:
I didn't get time to look into it in detail. At a high level, it seems fine. lgtm

ash211 commented:
Sounds good -- going ahead and merging

ash211 commented:
Oh uh oh, I merged instead of squashed and merged... which gave us a dirty history

going to:

0) temporarily enable force pushing to `branch-2.1-kubernetes`
1) force push to `branch-2.1-kubernetes` removing the merge commit
2) disable force pushing to `branch-2.1-kubernetes`
3) open a new PR from `mounted-secrets-auth` -> `branch-2.1-kubernetes`
4) **squash** and merge that PR

ash211 commented:
Done
